### ISSUE TITLE
WooCommerce post-type products

### ISSUE THREAD

**Issue by codingideas** [10up/ElasticPress#399]
How do we make this compatible with woocommerce? At this stage we have two searches:

?s=a&post_type=product
?s=b&post_type=post

?s=b&post_type=post works fine. We can see our posts. but if we want to ?s=a&post_type=product its not finding any products. 

What if we want to index taxonomies like pa_color&post_type=product which we use in the layered navigation. How would we do that?

**Comment by non-entity9** [comment]
I've literally just run into the same problem.

I can even see that the product information has been indexed, yet the search returns no results.

**Comment by codingideas** [comment]
Yes its weird because by default it says they search all taxonomies. Categories work. etc. Its just the search page with post_type=product. Shows no search results like you said.

**Comment by codingideas** [comment]
@tlovett1 do you know whats up with this?

**Comment by tlovett1** [comment]
I'm going to have to install Woocommerce to test. How the $_GET variable post_type is passed to WP_Query is probably somewhat dependent on your theme. Not 100% sure how it's setup. All public taxonomies are indexed but you have to use `search_fields` to designate which taxonomies to actually search.

**Comment by codingideas** [comment]
Thanks Taylor. Sure so in the search input we just add <input type="hidden"
name="post_type" value="product"> and we installed default woocommerce
products that come with woocommerce.

We've tried on multiple stock themes. We dont know if this is something we
need to ask for with custom code or if this is something that is suppose to
work.

**Comment by rafaelcmrj** [comment]
@tlovett1 As I see the query is trying to filter the post_meta._visibility but it isn't indexed.

See above the json for the filter param:

```
"filter":{  
      "and":[  
         {  
            "bool":{  
               "must":[  
                  {  
                     "terms":{  
                        "post_meta._visibility.raw":[  
                           "visible",
                           "search"
                        ]
                     }
                  }
               ]
            }
         },
         {  
            "term":{  
               "post_type.raw":"product"
            }
         }
      ]
   }
```

But the object isn't indexing post_meta._visibility, as you can see in the _source below:

```
"_source": {
                    "post_id": 93,
                    "post_author": {
                        "raw": "woo",
                        "login": "woo",
                        "display_name": "woo",
                        "id": 1
                    },
                    "post_date": "2013-06-07 11:36:34",
                    "post_date_gmt": "2013-06-07 11:36:34",
                    "post_title": "Woo Single #1",
                    "post_excerpt": "Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.",
                    "post_content": "<p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.</p>\n",
                    "post_status": "publish",
                    "post_name": "woo-single-1",
                    "post_modified": "2013-06-07 11:36:34",
                    "post_modified_gmt": "2013-06-07 11:36:34",
                    "post_parent": 0,
                    "post_type": "product",
                    "post_mime_type": "",
                    "permalink": "http://dev.elasticpress.com:8888/product/woo-single-1/",
                    "terms": {
                        "product_cat": [
                            {
                                "term_id": 11,
                                "slug": "music",
                                "name": "Music",
                                "parent": 0
                            },
                            {
                                "term_id": 13,
                                "slug": "singles",
                                "name": "Singles",
                                "parent": 11
                            }
                        ]
                    },
                    "post_meta": {
                        "total_sales": [
                            "0"
                        ]
                    },
                    "date_terms": {
                        "year": 2013,
                        "month": 6,
                        "week": 23,
                        "dayofyear": 157,
                        "day": 7,
                        "dayofweek": 5,
                        "dayofweek_iso": 5,
                        "hour": 11,
                        "minute": 36,
                        "second": 34,
                        "m": 201306
                    },
                    "comment_count": 0,
                    "comment_status": 0,
                    "ping_status": 0,
                    "menu_order": 0
                }
```
If I change "must" to "must_not" (or remove this filter) it works fine (filtering by the post_type).

**Comment by spaarndam** [comment]
When i comment out the following line in plugins/woocommerce/includes/class-wc-query.php.			
`add_action( 'pre_get_posts', array( $this, 'pre_get_posts' ) );` i'm able to search on the 'product' post type.

**Comment by tlovett1** [comment]
All, see https://github.com/10up/ElasticPress/blob/develop/classes/class-ep-api.php#L644

You need to allow the _visibility key to be index. Here is a snippet:

```php
add_filter( 'ep_prepare_meta_allowed_protected_keys', function( $keys ) {
    $keys[] = '_visibility';
    return $keys;
} );
```

**Comment by spaarndam** [comment]
Hi, thanks for the update.
But even with the snippet, there are still no results on the products.. 

 `add_filter( 'ep_prepare_meta_allowed_protected_keys', function( $keys ) {
		    $keys[] = '_visibility';
		    return $keys;
		}); ` 

 `$allowed_protected_keys = apply_filters( 'ep_prepare_meta_allowed_protected_keys', array(), $post );  `

**Comment by gabrielperezs** [comment]
Hello @spaarndam, the snippet works, but you have to reindex to use it. In my Woo instalation works perfect after the reindex.

Before the reindex, the product in elastichsearch didn't have this part:

                  "_visibility" : [
                     {
                        "value" : "visible",
                        "datetime" : "1971-01-01 00:00:01",
                        "raw" : "visible",
                        "boolean" : false,
                        "time" : "00:00:01",
                        "date" : "1971-01-01"
                     }
                  ],

**Comment by spaarndam** [comment]
Hi @gabrielperezs, 

yes, i managed to it working. Haven't thought about updating my comment here.. 
Problem for me, after reindexing was that ES indexed the products properly, but they still wouldn't come up during the search. 

After some checking, i found out that this was due to the way the woocommerce did the WP_Query.
I saw that woo did some checking on the visibility as well. 
I was able to get the results by overriding the woocommerce_product_query with the following snippet: 

 `function _woocommerce_product_query($query){
		 $query->query_vars['meta_query'][0] = null;
		 return $query;
}
add_filter('woocommerce_product_query','_woocommerce_product_query',1); `

**Comment by tlovett1** [comment]
Hey guys, this isn't officially released yet (in beta) but enjoy:
https://github.com/10up/elasticpress-woocommerce

**Comment by JosefJezek** [comment]
:+1:

**Comment by cwaynesmiley** [comment]
I know this thread is 2 years old, but I'm hoping to get some help.  I'm using WooCommerce 3.4.3 and Advanced Custom Fields.  For the default across the site I'm sorting the products by price high to low.  On a specific category (which has an ID of 87) I want to be able to search by the ACF relationship that I've established, which has a meta_key of bases_ordering.  Here's the code I've made so far but can't figure out why it's wrong or how to filter it properly.

`/* Custom Sorting for Checkout Process */
function my_pre_get_posts( $query ) {
	// do not modify queries in the admin
	if( is_admin() ) {
		return $query;
	}

	// only modify queries for 'event' post type
	if( isset($query->query_vars['post_type']) && $query->query_vars['post_type'] === 'product' && $query->query_vars['tag_ID'] === 87 ) {
		$query->set('orderby', 'meta_value');	
		$query->set('meta_key', 'bases_ordering');	 
		$query->set('order', 'DESC'); 
	}

	// return
	return $query;
}

add_action('pre_get_posts', 'my_pre_get_posts');`
